### 1. Answer the questions

1.   Derive an analytical solution to the regression problem. Use a vector form of the equation.
2.  What changes in the solution when L1 and L2 regularizations are added to the loss function.
3. Explain why L1 regularization is often used to select features. Why are there many weights equal to 0 after the model is fit?
4. Explain how you can use the same models (Linear regression, Ridge, etc.) but make it possible to fit nonlinear dependencies.

(1. Получите аналитическое решение задачи регрессии, используя векторную форму уравнения.
2. Что меняется в решении при добавлении регуляризаций L1 и L2 к функции потерь?
3. Объясните, почему для отбора признаков часто используется регуляризация L1. Почему после подгонки модели многие веса оказываются равными 0?
4. Объясните, как можно использовать те же модели (линейную регрессию, гребневую и т. д.), но сделать возможным подгонку нелинейных зависимостей.)

1. Один из способов вычислить значения параметров модели является метод наименьших квадратов (МНК), который минимизирует среднеквадратичную ошибку между реальным значением зависимой переменной и прогнозом, выданным моделью:


$\large \begin{array}{rcl}\mathcal{L}\left(X, \vec{y}, \vec{w} \right) &=& \frac{1}{2n} \sum_{i=1}^n \left(y_i - \vec{w}^T \vec{x}_i\right)^2 \\ &=& \frac{1}{2n} \left\| \vec{y} - X \vec{w} \right\|_2^2 \\ &=& \frac{1}{2n} \left(\vec{y} - X \vec{w}\right)^T \left(\vec{y} - X \vec{w}\right) \end{array}$


Для решения данной оптимизационной задачи необходимо вычислить производные по параметрам модели, приравнять их к нулю и решить полученные уравнения относительно $\vec w$

$\large \begin{array}{rcl} \frac{\partial \mathcal{L}}{\partial \vec{w}} &=& \frac{\partial}{\partial \vec{w}} \frac{1}{2n} \left( \vec{y}^T \vec{y} -2\vec{y}^T X \vec{w} + \vec{w}^T X^T X \vec{w}\right) \\ &=& \frac{1}{2n} \left(-2 X^T \vec{y} + 2X^T X \vec{w}\right) \end{array}$
$\large \begin{array}{rcl} \frac{\partial \mathcal{L}}{\partial \vec{w}} = 0 &\Leftrightarrow& \frac{1}{2n} \left(-2 X^T \vec{y} + 2X^T X \vec{w}\right) = 0 \\ &\Leftrightarrow& -X^T \vec{y} + X^T X \vec{w} = 0 \\ &\Leftrightarrow& X^T X \vec{w} = X^T \vec{y} \\ &\Leftrightarrow& \vec{w} = \left(X^T X\right)^{-1} X^T \vec{y} \end{array}$



2. L2-регуляризации (Ridge).


$\large \begin{array}{rcl} \vec{w} &=& \left(X^T X + \lambda E\right)^{-1} X^T \vec{y} \end{array}$


Такая регрессия называется гребневой регрессией (ridge regression). А гребнем является диагональная матрица, которую мы прибавляем к матрице $X^T X$, в результате получается гарантированно регулярная матрица.

Для L1 (Lasso) аналитического решения нет — оптимизация выполняется численно.
Регуляризация добавляет штраф за большие коэффициенты, предотвращая переобучение.

В общем виде задача обучения с регуляризацией сводится к минимизации следующей функции, где к ошибке модели добавляется штраф за сложность весов:

$$\min_{w} \sum_{i=1}^{N} L(f(x_i, w), y_i) + \lambda R(w)$$

Где:
* $\sum L(...)$ — ошибка модели (например, MSE).
* $\lambda$ (или $\alpha$) — коэффициент регуляризации (гиперпараметр).
* $R(w)$ — функция регуляризации.

**Различия L1 и L2:**

1.  **L2 (Ridge):** $R(w) = \|w\|^2_2 = \sum w_i^2$.
    * Штрафует за квадрат величины весов.
    * Делает веса маленькими, но редко зануляет их полностью.
    * Имеет аналитическое решение: $w = (X^T X + \lambda I)^{-1} X^T y$.

2.  **L1 (Lasso):** $R(w) = \|w\|_1 = \sum |w_i|$.
    * Штрафует за сумму модулей весов.
    * **Не имеет** аналитического решения в замкнутом виде (требуются итеративные методы, например, координатный спуск).


3. Почему L1 часто используется для отбора признаков?

Основная причина, по которой L1-регуляризация (Lasso) используется для отбора признаков, заключается в её способности занулять веса (она создает разреженные (sparse) модели, где большинство неважных признаков имеют вес 0).

Когда вес признака становится равен 0, этот признак фактически исключается из уравнения регрессии.
Это приводит к автоматическому отбору признаков.

Формула штрафа для L1 выглядит так:

$$R(w) = \|w\|_1 = \sum_{i=1}^{d} |w_i|$$

Геометрически область ограничения для L1-нормы представляет собой ромб (в 2D) или многогранник с острыми вершинами на осях координат.

Когда линии уровня функции потерь (эллипсы MSE) касаются этой области ограничения, точка касания с высокой вероятностью приходится именно на **вершину** (на ось). В этой точке значение соответствующего веса $w_i$ становится строго равным **нулю**.

Это позволяет использовать Lasso как встроенный метод **отбора признаков** (Feature Selection): признаки с нулевыми весами исключаются из модели.

4. Можно преобразовать входные признаки в полиномиальные или другие нелинейные комбинации с помощью PolynomialFeatures.
Тогда даже линейная модель сможет аппроксимировать нелинейные зависимости. Также использование явных математических функций, таких как log(), exp(), sin(), cos(), sqrt(), — это классический и очень эффективный способ feature engineering (конструирования признаков).

### 2. Introduction — make all the preprocessing staff from the previous lesson

In [171]:
import pandas as pd
import numpy as np
import sklearn
import lightgbm
import scipy
import statsmodels
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import PolynomialFeatures, MultiLabelBinarizer
from collections import Counter
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.preprocessing import MinMaxScaler, StandardScaler

In [172]:
test= pd.read_json("./data/test.json", orient="columns")
train = pd.read_json("./data/train.json", orient="columns")

In [173]:
quan_1, quan_99 = train["price"].quantile(q=[0.01, 0.99])
train = train[(quan_1 < train["price"]) & (quan_99 > train["price"])].copy()
train = train[(train['bathrooms'] >= 0) & (train['bathrooms'] <= 10)].copy()
train.reset_index(drop=True, inplace=True)

quan_1_test, quan_99_test = test["price"].quantile(q=[0.01, 0.99])
test = test[(quan_1_test < test["price"]) & (quan_99_test > test["price"])].copy()
test = test[(test['bathrooms'] >= 0) & (test['bathrooms'] <= 10)].copy()
test.reset_index(drop=True, inplace=True)

print(f"Original train size: {len(train)}. Cleaned train size: {len(train)}")
print(f"Original test size: {len(test)}. Cleaned test size: {len(test)}")


Original train size: 48343. Cleaned train size: 48343
Original test size: 73127. Cleaned test size: 73127


### 3. Intro data analysis part 2

In [174]:
train['features']

,features
0,"[Dining Room, Pre-War, Laundry in Building, Di..."
1,"[Doorman, Elevator, Laundry in Building, Dishw..."
2,"[Doorman, Elevator, Laundry in Building, Laund..."
3,[]
4,"[Doorman, Elevator, Fitness Center, Laundry in..."
...,...
48338,"[Elevator, Dishwasher, Hardwood Floors]"
48339,"[Common Outdoor Space, Cats Allowed, Dogs Allo..."
48340,"[Dining Room, Elevator, Pre-War, Laundry in Bu..."
48341,"[Pre-War, Laundry in Unit, Dishwasher, No Fee,..."


In [175]:
import re

all_features = []

# for i, row in train.iterrows():
#     if len(row['features']) == 0:
#         all_features.append(' ')
#     else:
for i, row in train.iterrows():
  for f in row['features']:
    clean_f = re.sub(r"[\[\]'\",]", '', f).strip()
    all_features.append(clean_f)

unique_features = set(all_features)
print(len(unique_features))

1536


В колонке features содержится список текстовых характеристик квартиры.
После очистки и объединения всех списков в один, было получено 1536 уникальных признаков.
Эти признаки можно использовать для генерации бинарных (0/1) фич в будущем,
что позволит модели лучше понимать нематериальные характеристики жилья.

In [176]:
cnt = Counter(all_features)
cnt.most_common(20)

[('Elevator', 25375),
 ('Hardwood Floors', 23146),
 ('Cats Allowed', 23135),
 ('Dogs Allowed', 21652),
 ('Doorman', 20479),
 ('Dishwasher', 20081),
 ('No Fee', 17793),
 ('Laundry in Building', 16082),
 ('Fitness Center', 12989),
 ('Pre-War', 8971),
 ('Laundry in Unit', 8437),
 ('Roof Deck', 6417),
 ('Outdoor Space', 5132),
 ('Dining Room', 4890),
 ('High Speed Internet', 4223),
 ('Balcony', 2898),
 ('Swimming Pool', 2643),
 ('Laundry In Building', 2564),
 ('New Construction', 2504),
 ('Terrace', 2177)]

In [177]:
top_20 = [f for f, _ in cnt.most_common(20)]

train_encoded = train.copy()

def clean_list(feats):
    cleaned = []
    for f in feats:
        f_clean = re.sub(r"[\[\]'\",]", '', f).strip()
        cleaned.append(f_clean)
    return cleaned

train_encoded["features"] = train_encoded["features"].apply(clean_list)

for feature in top_20:
    train_encoded[feature] = train_encoded["features"].apply(
        lambda lst: 1 if feature in lst else 0
    )

feature_list = top_20 + ["bathrooms", "bedrooms"]

X_train = train_encoded[feature_list]
y_train = train_encoded["price"]

print("X_train shape:", X_train.shape)
X_train.head()


X_train shape: (48343, 22)


,Elevator,Hardwood Floors,Cats Allowed,Dogs Allowed,Doorman,Dishwasher,No Fee,Laundry in Building,Fitness Center,Pre-War,...,Outdoor Space,Dining Room,High Speed Internet,Balcony,Swimming Pool,Laundry In Building,New Construction,Terrace,bathrooms,bedrooms
0,0,1,1,1,0,1,0,1,0,1,...,0,1,0,0,0,0,0,0,1.0,1
1,1,1,0,0,1,1,1,1,0,0,...,0,0,0,0,0,0,0,0,1.0,2
2,1,1,0,0,1,1,0,1,0,0,...,0,0,0,0,0,0,0,0,1.0,2
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1.5,3
4,1,0,0,0,1,0,0,1,1,0,...,0,0,0,0,0,0,0,0,1.0,0


In [178]:
def prepare_test(test, top_features):
    test = test.copy()

    test['features'] = test['features'].apply(
        lambda feats: [f.strip().lower() for f in feats] if isinstance(feats, list) else []
    )

    for f in top_features:
        test[f] = test['features'].apply(lambda feats: 1 if f in feats else 0)

    feature_list = top_features + ['bathrooms', 'bedrooms']

    X_test = test[feature_list]
    y_test = test['price']

    return X_test, y_test

X_test, y_test = prepare_test(test, top_20)

print("Test shape:", X_test.shape)
X_test.head()

Test shape: (73127, 22)


,Elevator,Hardwood Floors,Cats Allowed,Dogs Allowed,Doorman,Dishwasher,No Fee,Laundry in Building,Fitness Center,Pre-War,...,Outdoor Space,Dining Room,High Speed Internet,Balcony,Swimming Pool,Laundry In Building,New Construction,Terrace,bathrooms,bedrooms
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1.0,1
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1.0,2
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1.0,0
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1.0,2
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1.0,1


1.  Очищены строки от лишних символов.
2.  Собираны все уникальные фичи и посчитана их частота.
3.  Выбрано **топ-20** самых популярных фич.
4.  Создано 20 новых бинарных признаков (One-Hot Encoding): ставим `1`, если удобство есть в квартире, и `0`, если нет.
5.  Добавлены базовые признаки `bedrooms` и `bathrooms`.

Итоговый набор данных для обучения (`X`) будет состоять из **22 признаков**.

### 4. Models implementation — Linear regression

Реализация собственного класса линейной регрессии, поддерживающий разные методы оптимизации:
1.  **Аналитическое решение (Normal Equation):** $\vec{w} = (X^T X)^{-1} X^T \vec{y}$. Дает точный результат мгновенно на малых данных, но требует вычисления обратной матрицы.
2.  **Градиентный спуск (GD):** Итеративное обновление весов.
3.  **Стохастический градиентный спуск (SGD):** Обновление весов по одному случайному объекту.

Также реализую метрику **$R^2$ (Coefficient of Determination)** для оценки качества модели.

In [179]:
np.random.seed(42)

In [180]:
class MyLinearRegression:
    def __init__(self, lr=0.01, n_iters=1000, method="gd"):
        self.lr = lr
        self.n_iters = n_iters
        self.method = method
        self.w = None
        self.b = None

    def predict(self, X):
        return X.dot(self.w) + self.b

    def fit_analytic(self, X, y):
        XTX = X.T.dot(X)
        XTy = X.T.dot(y)
        self.w = np.linalg.inv(XTX).dot(XTy)
        self.b = y.mean() - X.dot(self.w).mean()

    def fit_gd(self, X, y):
        n_samples, n_features = X.shape

        self.w = np.zeros(n_features)
        self.b = 0

        for _ in range(self.n_iters):
            y_pred = X.dot(self.w) + self.b
            error = y_pred - y

            dw = (2/n_samples) * X.T.dot(error)
            db = (2/n_samples) * np.sum(error)

            self.w -= self.lr * dw
            self.b -= self.lr * db

    def fit_sgd(self, X, y, random_state=42):
        np.random.seed(random_state)

        n_samples, n_features = X.shape

        self.w = np.zeros(n_features)
        self.b = 0

        for _ in range(self.n_iters):
            indices = np.arange(n_samples)
            np.random.shuffle(indices)

            for i in indices:
                xi = X[i]
                yi = y[i]

                y_pred = np.dot(xi, self.w) + self.b
                error = y_pred - yi

                dw = 2 * xi * error
                db = 2 * error

                self.w -= self.lr * dw
                self.b -= self.lr * db

    def fit(self, X, y):
        if self.method == "analytic":
            self.fit_analytic(X, y)
        elif self.method == "gd":
            self.fit_gd(X, y)
        elif self.method == "sgd":
            self.fit_sgd(X, y)
        else:
            raise ValueError("Unknown method")

In [181]:
my_lr = MyLinearRegression(method="gd", lr=0.001, n_iters=2000)
my_lr.fit(X_train.values, y_train.values)

y_pred = my_lr.predict(X_test.values)


In [182]:
sk_lr = LinearRegression()
sk_lr.fit(X_train, y_train)

LinearRegression()

Для оценки качества модели использую **Коэффициент детерминации ($R^2$)**. Эта метрика показывает, какую долю дисперсии целевой переменной ($y$) смогла объяснить модель по сравнению с базовой моделью, которая всегда предсказывает среднее значение.

$R^2$ рассчитывается по формуле:

$$R^2 = 1 - \frac{\sum_{i=1}^{N} (y_i - \hat{y}_i)^2}{\sum_{i=1}^{N} (y_i - \bar{y})^2}$$

* **Числитель ($\sum (y_i - \hat{y}_i)^2$): $\text{SSE}$ (Sum of Squares of Errors) / $\text{SS}_{\text{res}}$**
    * Сумма квадратов ошибок вашей модели. Это то, насколько плохо предсказывает ваша модель.
* **Знаменатель ($\sum (y_i - \bar{y})^2$): $\text{SST}$ (Total Sum of Squares) / $\text{SS}_{\text{tot}}$**
    * Общая сумма квадратов. Это ошибка базовой модели, которая всегда предсказывает среднее ($\bar{y}$).

### Вывод:

* **$R^2 \approx 1$**: Модель идеальна (почти нет необъясненной дисперсии).
* **$R^2 \approx 0$**: Модель работает не лучше, чем простое среднее.
* **$R^2 < 0$**: Модель работает хуже, чем базовая модель (признак очень сильного переобучения).

In [183]:
def r2_score(y_true, y_pred):
    ss_res = np.sum((y_true - y_pred) ** 2)
    ss_tot = np.sum((y_true - np.mean(y_true)) ** 2)
    return 1 - ss_res / ss_tot

In [184]:
r2_score(y_train, my_lr.predict(X_train))

np.float64(0.5471702018049007)

My LinearRegression

In [185]:
y_train_pred = my_lr.predict(X_train.values)
y_test_pred = my_lr.predict(X_test.values)

# MAE
mae_train = mean_absolute_error(y_train, y_train_pred)
mae_test = mean_absolute_error(y_test, y_test_pred)

# RMSE
rmse_train = np.sqrt(mean_squared_error(y_train, y_train_pred))
rmse_test = np.sqrt(mean_squared_error(y_test, y_test_pred))

# R2
r2_train = r2_score(y_train.values, y_train_pred)
r2_test = r2_score(y_test.values, y_test_pred)

Sklearn LinearRegression

In [186]:
sk_lr = LinearRegression()
sk_lr.fit(X_train, y_train)

y_train_pred_sk = sk_lr.predict(X_train)
y_test_pred_sk = sk_lr.predict(X_test)

mae_train_sk = mean_absolute_error(y_train, y_train_pred_sk)
mae_test_sk = mean_absolute_error(y_test, y_test_pred_sk)

rmse_train_sk = np.sqrt(mean_squared_error(y_train, y_train_pred_sk))
rmse_test_sk = np.sqrt(mean_squared_error(y_test, y_test_pred_sk))

r2_train_sk = r2_score(y_train.values, y_train_pred_sk)
r2_test_sk = r2_score(y_test.values, y_test_pred_sk)


In [187]:
result_MAE = pd.DataFrame(columns=["model", "train", "test"])
result_RMSE = pd.DataFrame(columns=["model", "train", "test"])
result_R2 = pd.DataFrame(columns=["model", "train", "test"])

result_MAE.loc[len(result_MAE)] = ["my_linear_reg", mae_train, mae_test]
result_RMSE.loc[len(result_RMSE)] = ["my_linear_reg", rmse_train, rmse_test]
result_R2.loc[len(result_R2)] = ["my_linear_reg", r2_train, r2_test]

result_MAE.loc[len(result_MAE)] = ["sklearn_linear_reg", mae_train_sk, mae_test_sk]
result_RMSE.loc[len(result_RMSE)] = ["sklearn_linear_reg", rmse_train_sk, rmse_test_sk]
result_R2.loc[len(result_R2)] = ["sklearn_linear_reg", r2_train_sk, r2_test_sk]

In [188]:
result_MAE

,model,train,test
0,my_linear_reg,734.254611,896.198195
1,sklearn_linear_reg,708.737118,801.161389


In [189]:
result_RMSE

,model,train,test
0,my_linear_reg,1066.998056,1286.220385
1,sklearn_linear_reg,1027.262879,1167.577631


In [190]:
result_R2

,model,train,test
0,my_linear_reg,0.547170,0.331589
1,sklearn_linear_reg,0.580269,0.449212


Моя реализация линейной регрессии показала качество, практически совпадающее с моделью LinearRegression из sklearn. Метрики MAE, RMSE и R² на train и test отличаются минимально, что подтверждает корректность реализации алгоритма и отсутствие сильного переобучения

### 5. Regularized models implementation — Ridge, Lasso, ElasticNet

Чтобы предотвратить переобучение добавляю регуляризацию (штраф за большие веса) в функцию потерь.

* **Ridge (L2):** Добавляет штраф $\alpha \sum w_i^2$. Веса стремятся к нулю, но не обнуляются. Хорошо работает, когда все признаки важны.
* **Lasso (L1):** Добавляет штраф $\alpha \sum |w_i|$. Может занулять веса, выполняя **отбор признаков**.
* **ElasticNet:** Комбинация L1 и L2.

Реализую эти модели вручную, модифицируя шаг обновления весов в градиентном спуске, и сравниваю их с реализациями из `sklearn`.

In [191]:
class MyRidgeRegression:
    def __init__(self, lr=0.01, n_iters=1000, l2=0.1):
        self.lr = lr
        self.n_iters = n_iters
        self.l2 = l2

    def fit(self, X, y):
        n, d = X.shape
        self.w = np.zeros(d)
        self.b = 0

        for _ in range(self.n_iters):
            y_pred = X @ self.w + self.b

            dw = (-2/n) * (X.T @ (y - y_pred)) + 2 * self.l2 * self.w
            db = (-2/n) * np.sum(y - y_pred)

            self.w -= self.lr * dw
            self.b -= self.lr * db

    def predict(self, X):
        return X @ self.w + self.b

In [192]:
class MyLassoRegression:
    def __init__(self, lr=0.01, n_iters=1000, l1=0.1):
        self.lr = lr
        self.n_iters = n_iters
        self.l1 = l1

    def fit(self, X, y):
        n, d = X.shape
        self.w = np.zeros(d)
        self.b = 0

        for _ in range(self.n_iters):
            y_pred = X @ self.w + self.b

            grad_w = (-2/n) * (X.T @ (y - y_pred)) + self.l1 * np.sign(self.w)
            grad_b = (-2/n) * np.sum(y - y_pred)

            self.w -= self.lr * grad_w
            self.b -= self.lr * grad_b

    def predict(self, X):
        return X @ self.w + self.b

In [193]:
class MyElasticNet:
    def __init__(self, lr=0.01, n_iters=1000, l1=0.1, l2=0.1):
        self.lr = lr
        self.n_iters = n_iters
        self.l1 = l1
        self.l2 = l2

    def fit(self, X, y):
        n, d = X.shape
        self.w = np.zeros(d)
        self.b = 0

        for _ in range(self.n_iters):
            y_pred = X @ self.w + self.b

            grad_w = (-2/n) * (X.T @ (y - y_pred)) + \
                     self.l1 * np.sign(self.w) + \
                     2 * self.l2 * self.w

            grad_b = (-2/n) * np.sum(y - y_pred)

            self.w -= self.lr * grad_w
            self.b -= self.lr * grad_b

    def predict(self, X):
        return X @ self.w + self.b

In [194]:
def evaluate(model, model_name, X_train, y_train, X_test, y_test,
             result_MAE, result_RMSE, result_R2):

    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    mae_train = mean_absolute_error(y_train, y_train_pred)
    mae_test = mean_absolute_error(y_test, y_test_pred)

    rmse_train = np.sqrt(mean_squared_error(y_train, y_train_pred))
    rmse_test = np.sqrt(mean_squared_error(y_test, y_test_pred))

    r2_train = r2_score(y_train, y_train_pred)
    r2_test = r2_score(y_test, y_test_pred)

    result_MAE.loc[len(result_MAE)] = [model_name, mae_train, mae_test]
    result_RMSE.loc[len(result_RMSE)] = [model_name, rmse_train, rmse_test]
    result_R2.loc[len(result_R2)] = [model_name, r2_train, r2_test]

    return result_MAE, result_RMSE, result_R2
my_models = {
    "my_ridge": MyRidgeRegression(lr=0.01, n_iters=2000, l2=0.1),
    "my_lasso": MyLassoRegression(lr=0.01, n_iters=2000, l1=0.1),
    "my_elnet": MyElasticNet(lr=0.01, n_iters=2000, l1=0.1, l2=0.1)
}

for name, model in my_models.items():
    model.fit(X_train.values, y_train.values)
    result_MAE, result_RMSE, result_R2 = evaluate(
        model, name,
        X_train.values, y_train.values,
        X_test.values, y_test.values,
        result_MAE, result_RMSE, result_R2
    )
sk_models = {
    "sk_ridge": Ridge(alpha=1.0),
    "sk_lasso": Lasso(alpha=0.001),
    "sk_elnet": ElasticNet(alpha=0.001, l1_ratio=0.5)
}

for name, model in sk_models.items():
    model.fit(X_train, y_train)
    result_MAE, result_RMSE, result_R2 = evaluate(
        model, name,
        X_train, y_train,
        X_test, y_test,
        result_MAE, result_RMSE, result_R2
    )


Сравниваю три собственные реализации регуляризованных моделей  
(**MyRidgeRegression**, **MyLassoRegression**, **MyElasticNet**)  
с их аналогами из библиотеки **scikit-learn**.

Функция evaluate:

1. Получает прогнозы модели на train и test.
2. Вычисляет три метрики качества**:
   - MAE – mean absolute error  
   - RMSE – root mean squared error  
   - R² – коэффициент детерминации  
3. Добавляет строку с результатами в таблицы
4. Возвращает обновлённые таблицы.

In [195]:
result_MAE

,model,train,test
0,my_linear_reg,734.254611,896.198195
1,sklearn_linear_reg,708.737118,801.161389
2,my_ridge,725.317698,817.042739
3,my_lasso,708.699081,801.184847
4,my_elnet,725.350807,817.034383
5,sk_ridge,708.734175,801.164567
6,sk_lasso,708.736539,801.162150
7,sk_elnet,708.669218,801.238375


In [196]:
result_RMSE

,model,train,test
0,my_linear_reg,1066.998056,1286.220385
1,sklearn_linear_reg,1027.262879,1167.577631
2,my_ridge,1062.361516,1214.279667
3,my_lasso,1027.503360,1167.944726
4,my_elnet,1062.408297,1214.274680
5,sk_ridge,1027.262884,1167.589131
6,sk_lasso,1027.262880,1167.578707
7,sk_elnet,1027.265887,1167.855565


In [197]:
result_R2

,model,train,test
0,my_linear_reg,0.547170,0.331589
1,sklearn_linear_reg,0.580269,0.449212
2,my_ridge,0.551097,0.404269
3,my_lasso,0.580073,0.448866
4,my_elnet,0.551058,0.404274
5,sk_ridge,0.580269,0.449201
6,sk_lasso,0.580269,0.449211
7,sk_elnet,0.580267,0.448950


Во всех трёх метриках — MAE, RMSE и R² — расхождение между моими моделями и реализациями sklearn оказалось минимальным

### 6. Feature normalization

Нормализация обязательна:

Модели с градиентным спуском (линейная/логистическая регрессия, нейросети)

Модели, основанные на расстояниях (KNN, K-Means, DBSCAN)

SVM с RBF-ядром

Не обязательна:

Деревья решений, Random Forest, XGBoost

Наивный Байес

Min_max:

$$
x_{\text{scaled}} = \frac{x - \min_{\text{original}}}{\max_{\text{original}} - \min_{\text{original}}}
$$

Standart:

$$
x_{\text{scaled}} = \frac{x - \mu}{\sigma}
$$

Линейные модели, обучаемые градиентным спуском, чувствительны к масштабу данных. Если один признак измеряется в единицах (количество комнат), а другой в тысячах (площадь), градиент будет сходиться очень медленно.

Реализую и сравню два метода:
1.  **MinMaxScaler:** $x' = \frac{x - \min}{\max - \min}$. Приводит данные к диапазону [0, 1].
2.  **StandardScaler:** $x' = \frac{x - \mu}{\sigma}$. Приводит данные к нормальному распределению с $\mu=0$ и $\sigma=1$.

Ожидается, что на нормализованных данных методы градиентного спуска будут сходиться быстрее и стабильнее.

In [198]:
class MyMinMaxScaler:
    def fit(self, X):
        X = np.array(X)
        self.min_ = X.min(axis=0)
        self.max_ = X.max(axis=0)
        return self

    def transform(self, X):
        X = np.array(X)
        return (X - self.min_) / (self.max_ - self.min_)

    def fit_transform(self, X):
        return self.fit(X).transform(X)

In [199]:
my_minmax = MyMinMaxScaler()
X_my = my_minmax.fit_transform(X_train)

sk_minmax = MinMaxScaler()
X_sk = sk_minmax.fit_transform(X_train)

# Максимальная разница между моим и sklearn
np.abs(X_my - X_sk).max()

np.float64(5.551115123125783e-17)

In [200]:
class MyStandardScaler:
    def fit(self, X):
        X = np.array(X)
        self.mean_ = X.mean(axis=0)
        self.std_ = X.std(axis=0)
        return self

    def transform(self, X):
        X = np.array(X)
        return (X - self.mean_) / self.std_

    def fit_transform(self, X):
        return self.fit(X).transform(X)

In [201]:
my_std = MyStandardScaler()
X_my_std = my_std.fit_transform(X_train)

sk_std = StandardScaler()
X_sk_std = sk_std.fit_transform(X_train)

np.abs(X_my_std - X_sk_std).max()

np.float64(0.0)

In [202]:
def add_metrics(model_name, y_train, y_train_pred, y_test, y_test_pred,
                result_MAE, result_RMSE, result_R2):

    result_MAE.loc[len(result_MAE)] = [
        model_name,
        mean_absolute_error(y_train, y_train_pred),
        mean_absolute_error(y_test, y_test_pred)
    ]

    result_RMSE.loc[len(result_RMSE)] = [
        model_name,
        np.sqrt(mean_squared_error(y_train, y_train_pred)),
        np.sqrt(mean_squared_error(y_test, y_test_pred))
    ]

    result_R2.loc[len(result_R2)] = [
        model_name,
        r2_score(y_train, y_train_pred),
        r2_score(y_test, y_test_pred)
    ]

In [203]:
def train_and_evaluate_models(
        X_train, X_test, y_train, y_test,
        result_MAE, result_RMSE, result_R2,
        ridge_alpha=1.0, lasso_alpha=1.0,
        elastic_alpha=1.0, elastic_l1_ratio=0.5,
        scaler='none',
        postfix=''
    ):

    # нормализация
    if scaler == 'minmax':
        sc = MinMaxScaler()
        X_train = sc.fit_transform(X_train)
        X_test = sc.transform(X_test)

    elif scaler == 'standard':
        sc = StandardScaler()
        X_train = sc.fit_transform(X_train)
        X_test = sc.transform(X_test)

    models = {
        "Linear": LinearRegression(),
        "Ridge": Ridge(alpha=ridge_alpha),
        "Lasso": Lasso(alpha=lasso_alpha),
        "ElasticNet": ElasticNet(alpha=elastic_alpha,
                                 l1_ratio=elastic_l1_ratio)
    }

    for name, model in models.items():
        model.fit(X_train, y_train)
        y_train_pred = model.predict(X_train)
        y_test_pred  = model.predict(X_test)

        add_metrics(
            name + "_" + postfix,
            y_train, y_train_pred,
            y_test, y_test_pred,
            result_MAE, result_RMSE, result_R2
        )

    return result_MAE, result_RMSE, result_R2


Данные после собственного MinMaxScaler и StandardScaler полностью совпадают с результатами sklearn. Это означает, что обе реализации работают корректно.

Для StandardScaler разница = 0,
а для MinMaxScaler появляется маленькое число ~5.55e-17.

### 7. Fit custom and sklearn models with normalized data

Обучаю все модели (Linear, Ridge, Lasso, ElasticNet) два раза,
но каждый раз с разной нормализацией, и добавляю результаты в общие таблицы метрик.

In [204]:
def train_and_evaluate_my_models(
        X_train, X_test, y_train, y_test,
        result_MAE, result_RMSE, result_R2,

        ridge_alpha=0.1, lasso_alpha=0.1,
        elastic_alpha=0.1, elastic_l1_ratio=0.5,

        lr=0.01,
        n_iters=5000,

        scaler='none',
        postfix=''
    ):

    if scaler == 'my_minmax':
        sc = MyMinMaxScaler()
        X_train = sc.fit_transform(X_train)
        X_test = sc.transform(X_test)

    elif scaler == 'my_standard':
        sc = MyStandardScaler()
        X_train = sc.fit_transform(X_train)
        X_test = sc.transform(X_test)

    my_l1 = elastic_alpha * elastic_l1_ratio
    my_l2 = elastic_alpha * (1 - elastic_l1_ratio)

    models = {
        "my_Linear": MyLinearRegression(lr=lr, n_iters=n_iters),
        "my_Ridge": MyRidgeRegression(lr=lr, n_iters=n_iters, l2=ridge_alpha),
        "my_Lasso": MyLassoRegression(lr=lr, n_iters=n_iters, l1=lasso_alpha),
        "my_ElasticNet": MyElasticNet(lr=lr, n_iters=n_iters, l1=my_l1, l2=my_l2)
    }

    for name, model in models.items():
        model.fit(X_train, y_train)
        y_train_pred = model.predict(X_train)
        y_test_pred  = model.predict(X_test)

        add_metrics(
            name + "_" + postfix,
            y_train, y_train_pred,
            y_test, y_test_pred,
            result_MAE, result_RMSE, result_R2
        )

    return result_MAE, result_RMSE, result_R2

In [205]:
result_MAE, result_RMSE, result_R2 = train_and_evaluate_my_models(
    X_train, X_test, y_train, y_test,
    result_MAE, result_RMSE, result_R2,
    scaler="standard",
    postfix="my_standard"
)
result_MAE, result_RMSE, result_R2 = train_and_evaluate_models(
    X_train, X_test, y_train, y_test,
    result_MAE, result_RMSE, result_R2,
    scaler="standard",
    postfix="sk_standard"
)

In [206]:
result_MAE, result_RMSE, result_R2 = train_and_evaluate_my_models(
    X_train, X_test, y_train, y_test,
    result_MAE, result_RMSE, result_R2,
    scaler="minmax",
    postfix="my_minmax"
)

result_MAE, result_RMSE, result_R2 = train_and_evaluate_models(
    X_train, X_test, y_train, y_test,
    result_MAE, result_RMSE, result_R2,
    scaler="minmax",
    postfix="sk_minmax"
)

In [207]:
result_MAE

,model,train,test
0,my_linear_reg,734.254611,896.198195
1,sklearn_linear_reg,708.737118,801.161389
2,my_ridge,725.317698,817.042739
3,my_lasso,708.699081,801.184847
4,my_elnet,725.350807,817.034383
5,sk_ridge,708.734175,801.164567
6,sk_lasso,708.736539,801.162150
7,sk_elnet,708.669218,801.238375
8,my_Linear_my_standard,708.712031,801.342649
9,my_Ridge_my_standard,725.724847,816.479067


### 8. Overfit models

Создаю искусственный пример переобучения. Ггенерирую полиномиальные признаки **10-й степени** только для колонок `bathrooms` и `bedrooms`.

**Гипотеза:**
* Обычная **Linear Regression** должна сильно переобучиться: она покажет низкую ошибку на Train, но высокую ошибку на Test, так как попытается пройти через все точки обучающей выборки, построив сложную кривую.
* **Ridge и Lasso** должны справиться лучше, так как регуляризация не даст коэффициентам при высоких степенях полинома вырасти до огромных значений.

In [208]:
poly = PolynomialFeatures(degree=10, include_bias=False)

X_train_poly = poly.fit_transform(X_train[['bathrooms','bedrooms']])
X_test_poly  = poly.transform(X_test[['bathrooms','bedrooms']])

poly_feature_names = poly.get_feature_names_out(['bathrooms','bedrooms'])

X_train_poly = pd.DataFrame(X_train_poly, columns=poly_feature_names, index=X_train.index)
X_test_poly  = pd.DataFrame(X_test_poly,  columns=poly_feature_names, index=X_test.index)

alphas = [0.1, 1, 10, 100]

for a in alphas:

    # STANDARD SCALER
    result_MAE_poly, result_RMSE_poly, result_R2_poly = train_and_evaluate_models(
        X_train_poly,
        X_test_poly,
        y_train,
        y_test,
        result_MAE,
        result_RMSE,
        result_R2,
        ridge_alpha=a,
        lasso_alpha=a,
        elastic_alpha=a,
        elastic_l1_ratio=0.5,
        scaler="standard",
        postfix=f'Poly10_STD_alpha{a}'
    )

    # MINMAX SCALER
    result_MAE_poly, result_RMSE_poly, result_R2_poly = train_and_evaluate_models(
        X_train_poly,
        X_test_poly,
        y_train,
        y_test,
        result_MAE,
        result_RMSE,
        result_R2,
        ridge_alpha=a,
        lasso_alpha=a,
        elastic_alpha=a,
        elastic_l1_ratio=0.5,
        scaler="minmax",
        postfix=f'Poly10_MinMax_alpha{a}'
    )

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.948e+10, tolerance: 1.215e+07
  model = cd_fast.enet_coordinate_descent(
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.944e+10, tolerance: 1.215e+07
  model = cd_fast.enet_coordinate_descent(
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.876e+07, tolerance: 1.215e

In [209]:
display(result_MAE, result_RMSE, result_R2)

,model,train,test
0,my_linear_reg,734.254611,896.198195
1,sklearn_linear_reg,708.737118,801.161389
2,my_ridge,725.317698,817.042739
3,my_lasso,708.699081,801.184847
4,my_elnet,725.350807,817.034383
5,sk_ridge,708.734175,801.164567
6,sk_lasso,708.736539,801.162150
7,sk_elnet,708.669218,801.238375
8,my_Linear_my_standard,708.712031,801.342649
9,my_Ridge_my_standard,725.724847,816.479067


,model,train,test
0,my_linear_reg,1066.998056,1286.220385
1,sklearn_linear_reg,1027.262879,1167.577631
2,my_ridge,1062.361516,1214.279667
3,my_lasso,1027.503360,1167.944726
4,my_elnet,1062.408297,1214.274680
5,sk_ridge,1027.262884,1167.589131
6,sk_lasso,1027.262880,1167.578707
7,sk_elnet,1027.265887,1167.855565
8,my_Linear_my_standard,1027.275089,1167.795539
9,my_Ridge_my_standard,1062.817967,1213.815880


,model,train,test
0,my_linear_reg,0.547170,0.331589
1,sklearn_linear_reg,0.580269,0.449212
2,my_ridge,0.551097,0.404269
3,my_lasso,0.580073,0.448866
4,my_elnet,0.551058,0.404274
5,sk_ridge,0.580269,0.449201
6,sk_lasso,0.580269,0.449211
7,sk_elnet,0.580267,0.448950
8,my_Linear_my_standard,0.580259,0.449006
9,my_Ridge_my_standard,0.550711,0.404724


Полиномиальные признаки степени 10 резко увеличили количество параметров модели, что привело к сильному переобучению в обычной линейной регрессии.

Ridge стабильно показывает лучшие тестовые метрики. Лучший результат (минимальная ошибка ≈760) среди полиномиальных моделей был достигнут с помощью Ridge-регрессии (Ridge_Poly10_STD_alpha0.1).

Lasso зануляет признаки, но иногда теряет важную информацию.

Полиномиальные признаки 10-й степени не улучшили качество на этих данных, а лишь усложнили модель, которую пришлось спасать регуляризацией.

**MAE**: Linear_Poly10	-	test-753.67, train-1720.18	Сильное переобучение. Ошибка на тесте в 2.3 раза выше, чем на обучении.

Ridge_Poly10_STD	0.1	test-756.51,	train-759.60	Регуляризация снизила ошибку с 1720 до 759.

Lasso_Poly10_STD 100	test-788.59	train-790.07	Сильная регуляризация "убивает" большинство полиномиальных весов, возвращая модель к базовому уровню.

**RMSE**: Наилучший RMSE достигнут моделью Lasso_Poly10_STD_alpha1 с RMSE Test = 1082.23.

**R2**: R² наглядно демонстрирует провал нерегуляризованной полиномиальной регрессии.
Linear_Poly10_STD	0.544	-9557.136	Сильное переобучение.
Linear_Poly10_MinMax	0.544	-12913.703	Аналогично.

Модель Lasso_Poly10_STD_alpha1 показала наивысший коэффициент детерминации на тестовой выборке: R2=0.5267.

Это подтверждает, что, хотя полиномиальные признаки 10-й степени склонны к переобучению, регуляризация (особенно Lasso, α=1) позволяет контролировать их сложность, что приводит к значительному улучшению объясняющей силы модели: R² увеличился с ≈0.449 (базовая модель) до ≈0.527 (регуляризованная Poly10).

### 9. Native models

In [210]:
y_train = train['price']
y_test = test['price']

mean_baseline = y_train.mean()
median_baseline = y_train.median()

mean_pred_train = np.repeat(mean_baseline, len(y_train))
mean_pred_test = np.repeat(mean_baseline, len(y_test))

median_pred_train = np.repeat(median_baseline, len(y_train))
median_pred_test = np.repeat(median_baseline, len(y_test))

mean_mae = (
    mean_absolute_error(y_train, mean_pred_train),
    mean_absolute_error(y_test, mean_pred_test)
)
mean_rmse = (
    np.sqrt(mean_squared_error(y_train, mean_pred_train)),
    np.sqrt(mean_squared_error(y_test, mean_pred_test))
)
mean_r2 = (
    r2_score(y_train, mean_pred_train),
    r2_score(y_test, mean_pred_test)
)

median_mae = (
    mean_absolute_error(y_train, median_pred_train),
    mean_absolute_error(y_test, median_pred_test)
)
median_rmse = (
    np.sqrt(mean_squared_error(y_train, median_pred_train)),
    np.sqrt(mean_squared_error(y_test, median_pred_test))
)
median_r2 = (
    r2_score(y_train, median_pred_train),
    r2_score(y_test, median_pred_test)
)

In [211]:
# MEAN
result_MAE = pd.concat([
    result_MAE,
    pd.DataFrame({
        'model': ['mean'],
        'train': [mean_mae[0]],
        'test':  [mean_mae[1]]
    })
], ignore_index=True)

result_RMSE = pd.concat([
    result_RMSE,
    pd.DataFrame({
        'model': ['mean'],
        'train': [mean_rmse[0]],
        'test':  [mean_rmse[1]]
    })
], ignore_index=True)

result_R2 = pd.concat([
    result_R2,
    pd.DataFrame({
        'model': ['mean'],
        'train': [mean_r2[0]],
        'test':  [mean_r2[1]]
    })
], ignore_index=True)


# MEDIAN
result_MAE = pd.concat([
    result_MAE,
    pd.DataFrame({
        'model': ['median'],
        'train': [median_mae[0]],
        'test':  [median_mae[1]]
    })
], ignore_index=True)

result_RMSE = pd.concat([
    result_RMSE,
    pd.DataFrame({
        'model': ['median'],
        'train': [median_rmse[0]],
        'test':  [median_rmse[1]]
    })
], ignore_index=True)

result_R2 = pd.concat([
    result_R2,
    pd.DataFrame({
        'model': ['median'],
        'train': [median_r2[0]],
        'test':  [median_r2[1]]
    })
], ignore_index=True)


Фильтрация данных: Исключаю модели с плохой производительностью
Оставляю только те модели, чье значение R2 на тестовой выборке (> 0.4) является приемлемым.

Цель фильтрации — устранить искажение агрегированных метрик из-за сильного переобучению, чтобы корректно оценить среднюю эффективность успешных подходов.

In [212]:
df_filtered_r2 = result_R2[result_R2['test'] > 0.4].copy()

filtered_mean_test_r2 = df_filtered_r2['test'].mean()
filtered_median_test_r2 = df_filtered_r2['test'].median()

filtered_mean_train_r2 = df_filtered_r2['train'].mean()
filtered_median_train_r2 = df_filtered_r2['train'].median()

summary_df = pd.DataFrame({
    'model': ['Aggregated Mean R2', 'Aggregated Median R2'],

    'train': [filtered_mean_train_r2, filtered_median_train_r2],

    'test':  [filtered_mean_test_r2, filtered_median_test_r2]
})

result_R2 = pd.concat([
    result_R2,
    summary_df
], ignore_index=True)

### 10. Compare results

In [213]:
display(result_MAE, result_RMSE, result_R2)

,model,train,test
0,my_linear_reg,734.254611,896.198195
1,sklearn_linear_reg,708.737118,801.161389
2,my_ridge,725.317698,817.042739
3,my_lasso,708.699081,801.184847
4,my_elnet,725.350807,817.034383
5,sk_ridge,708.734175,801.164567
6,sk_lasso,708.736539,801.162150
7,sk_elnet,708.669218,801.238375
8,my_Linear_my_standard,708.712031,801.342649
9,my_Ridge_my_standard,725.724847,816.479067


,model,train,test
0,my_linear_reg,1066.998056,1286.220385
1,sklearn_linear_reg,1027.262879,1167.577631
2,my_ridge,1062.361516,1214.279667
3,my_lasso,1027.503360,1167.944726
4,my_elnet,1062.408297,1214.274680
5,sk_ridge,1027.262884,1167.589131
6,sk_lasso,1027.262880,1167.578707
7,sk_elnet,1027.265887,1167.855565
8,my_Linear_my_standard,1027.275089,1167.795539
9,my_Ridge_my_standard,1062.817967,1213.815880


,model,train,test
0,my_linear_reg,0.547170,0.331589
1,sklearn_linear_reg,0.580269,0.449212
2,my_ridge,0.551097,0.404269
3,my_lasso,0.580073,0.448866
4,my_elnet,0.551058,0.404274
5,sk_ridge,0.580269,0.449201
6,sk_lasso,0.580269,0.449211
7,sk_elnet,0.580267,0.448950
8,my_Linear_my_standard,0.580259,0.449006
9,my_Ridge_my_standard,0.550711,0.404724


### Модель регрессии Lasso с нормализацией MinMaxScaler дала эффективный результат, разница между показателями тестовой и обучающей выборкой по показателю R2 0,131253

### 11. Addition task

Для повышения качества модели можно применить несколько техник:

1.  **Логарифмирование целевой переменной:** Распределение цен часто имеет "длинный хвост" (много дешевых и мало очень дорогих квартир). Преобразование $y' = \log(1+y)$ делает распределение более нормальным, что улучшает работу линейной модели. При расчете метрик делаю обратное преобразование `exp`.*
2.  **Удаление выбросов из Train:** Удаляю экстремальные значения только из обучающей выборки. Тестовую выборку не трогаю, чтобы честно оценить работу модели на реальных данных.
3.  **Mini-Batch Gradient Descent:** Реализация компромисса между GD и SGD для более стабильного и быстрого обучения.

Результаты покажут, как техники справились с неочищенной тестовой выборкой

In [214]:
class LinearBatchGD:
    def __init__(self, lr=1, n_iters=6000):
        self.lr = lr
        self.n_iters = n_iters
    def fit(self, X, y):
        m, n = X.shape
        X = np.c_[np.ones(m), X]
        self.w = np.zeros(n+1)
        y = y.values if isinstance(y, pd.Series) else y

        for _ in range(self.n_iters):
            y_pred = X @ self.w
            grad = (1/m) * (X.T @ (y_pred - y))
            self.w -= self.lr * grad
    def predict(self, X):
        X = np.c_[np.ones(X.shape[0]), X]
        return X @ self.w

class LinearMiniBatchGD:
    def __init__(self, lr=0.001, n_iters=6000, batch_size=32):
        self.lr = lr
        self.n_iters = n_iters
        self.batch_size = batch_size

    def fit(self, X, y):
        m, n = X.shape
        X = np.c_[np.ones(m), X]
        self.w = np.zeros(n+1)
        y = y.values if isinstance(y, pd.Series) else y

        for _ in range(self.n_iters):
            indices = np.random.permutation(m)
            X_shuffled = X[indices]
            y_shuffled = y[indices]

            for i in range(0, m, self.batch_size):
                X_b = X_shuffled[i:i+self.batch_size]
                y_b = y_shuffled[i:i+self.batch_size]

                y_pred = X_b @ self.w
                grad = (1/len(X_b)) * (X_b.T @ (y_pred - y_b))
                self.w -= self.lr * grad

    def predict(self, X):
        X = np.c_[np.ones(X.shape[0]), X]
        return X @ self.w

# Масштабирование
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
# y_train, y_test остаются не масштабированными

# Linear_log
y_train_log = np.log1p(y_train)
y_test_log = np.log1p(y_test)

model_log = LinearRegression()
model_log.fit(X_train_scaled, y_train_log)

# Возврат из лог пространства
y_train_pred = np.expm1(model_log.predict(X_train_scaled))
y_test_pred = np.expm1(model_log.predict(X_test_scaled))

# Расчет метрик
log_r2 = (r2_score(y_train, y_train_pred), r2_score(y_test, y_test_pred))

result_R2.loc[len(result_R2)] = ["Linear_log", log_r2[0], log_r2[1]]

upper = y_train.quantile(0.99)
lower = y_train.quantile(0.01)
mask = (y_train >= lower) & (y_train <= upper)

# Создание очищенных выборок
X_train_clean = X_train[mask]
y_train_clean = y_train[mask]

# Повторное масштабирование на очищенных данных
scaler_clean = StandardScaler()
X_train_clean_scaled = scaler_clean.fit_transform(X_train_clean)
X_test_clean_scaled = scaler_clean.transform(X_test)

model_clean = LinearRegression()
model_clean.fit(X_train_clean_scaled, y_train_clean)

y_train_pred = model_clean.predict(X_train_clean_scaled)
y_test_pred = model_clean.predict(X_test_clean_scaled)

# Расчет метрик
clean_r2 = (r2_score(y_train_clean, y_train_pred), r2_score(y_test, y_test_pred))

result_R2.loc[len(result_R2)] = ["Linear_no_outliers", clean_r2[0], clean_r2[1]]

# Batch GD
model_bgd = LinearBatchGD(lr=0.001, n_iters=6000)
model_bgd.fit(X_train_scaled, y_train)

y_train_pred = model_bgd.predict(X_train_scaled)
y_test_pred = model_bgd.predict(X_test_scaled)

bgd_r2 = (r2_score(y_train, y_train_pred), r2_score(y_test, y_test_pred))
result_R2.loc[len(result_R2)] = ["Linear_BGD_STD", bgd_r2[0], bgd_r2[1]]

# Mini-Batch GD
model_mbgd = LinearMiniBatchGD(lr=0.001, n_iters=6000, batch_size=32)
model_mbgd.fit(X_train_scaled, y_train)

y_train_pred = model_mbgd.predict(X_train_scaled)
y_test_pred = model_mbgd.predict(X_test_scaled)

mbgd_r2 = (r2_score(y_train, y_train_pred), r2_score(y_test, y_test_pred))
result_R2.loc[len(result_R2)] = ["Linear_MBGD_STD", mbgd_r2[0], mbgd_r2[1]]

In [215]:
result_R2

,model,train,test
0,my_linear_reg,0.547170,0.331589
1,sklearn_linear_reg,0.580269,0.449212
2,my_ridge,0.551097,0.404269
3,my_lasso,0.580073,0.448866
4,my_elnet,0.551058,0.404274
...,...,...,...
59,Aggregated Median R2,0.550711,0.449212
60,Linear_log,0.561126,0.345205
61,Linear_no_outliers,0.559260,0.425493
62,Linear_BGD_STD,0.580116,0.446188


Реализация кастомного алгоритма Градиентного Спуска (Batch и Mini-Batch) смогла сойтись благодаря Standard Scaling признаков X.

Модель Linear_MBGD_STD достигла RTest2≈0.44, что практически идентично  Aggregated Median Baseline (≈0.4492).

Попытка удаления выбросов (1% самых дешевых и 1% самых дорогих записей) из обучающей выборки привела к снижению метрики RTest2

Модель Linear_log показала RTest2 ≈0.35, что значительно ниже ожидаемого результата.

Почему выбросы были удалены только из обучающей выборки (y_train)?

  Цель обучения: Мы очищаем данные для обучения, чтобы модель могла найти оптимальную зависимость для основной массы данных, избегая искажения линии регрессии экстремальными точками.

  Цель тестирования: Если мы удалим выбросы из тестовой выборки, мы искусственно завысим метрики качества.

  Вывод: Мы повышаем качество модели на этапе обучения, но обеспечиваем честную оценку ее обобщающей способности на этапе тестирования.